pkg/flow/internal/controller/loader.go

```diff
@@ -154,6 +153,34 @@ func (l *Loader) Apply(args map[string]any, componentBlocks []*ast.BlockStmt, co
 
 	l.cache.ClearModuleExports()
 
+	// Prioritize evaluate the logging block to ensure logging format conformity
+	for _, leave := range newGraph.Leaves() {
+		if leave.NodeID() != "logging" {
+			continue
+		}
+
+		n := leave.(BlockNode)
+
+		_, span := tracer.Start(spanCtx, "EvaluateNode", trace.WithSpanKind(trace.SpanKindInternal))
+		span.SetAttributes(attribute.String("node_id", n.NodeID()))
+		if err := l.evaluate(logger, n); err != nil {
+			diags.Add(diag.Diagnostic{
+				Severity: diag.SeverityLevelError,
+				Message:  fmt.Sprintf("Failed to evaluate node for config block: %s", err),
+				StartPos: ast.StartPos(n.Block()).Position(),
+				EndPos:   ast.EndPos(n.Block()).Position(),
+			})
+			span.SetStatus(codes.Error, err.Error())
+		}
+		if exp, ok := n.(*ExportConfigNode); ok {
+			l.cache.CacheModuleExportValue(exp.Label(), exp.Value())
+		}
+		level.Info(logger).Log("msg", "finished node evaluation", "node_id", n.NodeID(), "duration", time.Since(start))
+		span.SetStatus(codes.Ok, "")
+	}
```

Comment:
Hey, thanks for contributing again :) 
The problem with this approach is that the logging node might not be a leaf. It might actually depend on another node like in this config:

```
local.file "LABEL" {
  filename = "tmp.txt"
}

logging {
  level  = "info"
  format = local.file.LABEL.content
}
```

In this case you would first need to evaluate the local.file node before being able to evaluate the logging node, which would produce some logs in the wrong format.

Reply:
Oh shoot, thats the edge case I havent thought about. Maybe we can apply the evaluation for leaf-only logging node?

I think for logging cfg, most ppl would tend to be explicit. But thats my naive assumption 😁

Reply:
Maybe there is a way to buffer the logs till the logging node is evaluated so that they can be converted in the right format before being written? @rfratto what do you think?

Reply:
IMO, I think buffering logs until the first evaluation of the block will likely be easier to understand and maintain long term.

Reply:
Hi folks, when I'm implementing the buffered logger, I just discovered that there are components using modules that also emits logs and we need to buffer them into account. If we decide to tackle this, the scope will vastly increase

For example, [`prometheus.remote_write` component with its usage of `go-kit/level/log`](https://github.com/prometheus/prometheus/blob/6150e1ca0ede508e56414363cc9062ef522db518/tsdb/wlog/watcher.go#L222)

<img width="1111" alt="image" src="https://github.com/grafana/agent/assets/41283691/84f233de-17e6-48ef-ab9e-d797b103834b">

I think my PR and a little bit of guidance, i.e. informing folks to explicitly declare `logging` node, would be fine to resolve to this. 

Wdyt?

Reply:
Sure, it's a small one

```
logging {
  format = "json"
}

module.git "test" {
  repository = "https://github.com/ntimo/grafana-agent-git-module-test"
  path       = "prometheus.river"
}
```

Reply:
Thanks, I had a quick look at it, in logger.go, all the logs are passing through `Log(kvps ...interface{})`. There it's possible to store the logs in a buffer instead of passing them to `slogadapter.GoKit(l.handler).Log(kvps...)`. The Update function is called twice, once on New and once on Evaluate. I think that we could actually not run it in the New function. That way, when the first Update comes you can empty the buffer and stop intercepting the calls in the `Log(kvps ...interface{})` function. I'm out of office now but I can help with it beginning next week if needed.

Reply:
@hainenber Can you share what you have for the buffering approach? I'd like to see what it's looking like and identify whether I can help get it finished and merged.

Reply:
Hi there, I've got a busy schedule last week. I made a draft [PR](https://github.com/hainenber/agent/pull/1) here for the buffering approach I was working on previously.

Reply:
Thanks for taking a shot at it :) I don't believe that the code you shared in your PR is the right approach. I think that the problem can be solved by adding the buffering inside of the pkg/flow/logging/logger.go with the solution I described above. Only a few changes within this file should be enough to solve the issue.
Please don't feel pressured, if you have too much on your plate I'd be happy to take over this issue. But if you want to continue working on it and the approach I described is unclear don't hesitate to ask questions.

Reply:
hi there, I'll give another shot using your suggestion, thanks!

Reply:
Holy moly, your suggestion works @wildum! 

<img width="1116" alt="image" src="https://github.com/grafana/agent/assets/41283691/1d0c75cf-3dd0-487c-ae41-81c9b92883a3">